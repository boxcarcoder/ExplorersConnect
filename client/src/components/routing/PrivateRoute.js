import React from 'react';
import PropTypes from 'prop-types';
import { Route, Redirect } from 'react-router-dom';

//redux
import { connect } from 'react-redux';

const PrivateRoute = ({
  component: Component,
  authState: { isAuthenticated, loading },
  ...rest //any custom props that are passed in
}) => (
  <Route
    {...rest}
    render={props =>
      !isAuthenticated && !loading ? (
        <Redirect to='/login' />
      ) : (
        <Component {...props} />
      )
    }
  />
);

PrivateRoute.propTypes = {
  authState: PropTypes.object.isRequired
};

const mapStateToProps = state => ({
  authState: state.auth
});

export default connect(mapStateToProps)(PrivateRoute);
