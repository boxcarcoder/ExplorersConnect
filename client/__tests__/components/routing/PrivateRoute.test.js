import React from 'react';
import {  mount } from 'enzyme';
import { MemoryRouter } from 'react-router';

// Import the non-connected component
import { PrivateRoute } from '../../../src/components/routing/PrivateRoute';

// Globals for testing
let wrapper;
let props;
let mockComponent = () => <div>Component</div> ;

describe('<PrivateRoute /> component.', () => {
    describe('Successfully renders', () => {
        test('the component.', () => {
            props = {
                component: mockComponent,
                authState: {
                    isAuthenticated: true,
                    loading: false
                }
            }

            wrapper = mount(
                <MemoryRouter> 
                    <PrivateRoute {...props} />
                </MemoryRouter>);
            expect(wrapper).toMatchSnapshot();
        });

        test('a Redirect to the login page if isAuthenticated and loading are false in the Auth redux state.', () => {
            props = {
                component: mockComponent,
                authState: {
                    isAuthenticated: false,
                    loading: false
                }
            }

            wrapper = mount(
                <MemoryRouter> 
                    <PrivateRoute {...props} />
                </MemoryRouter>);

            const history = wrapper.find('Router').prop('history');
            expect(history.location.pathname).toBe('/login');
        });

    });

});