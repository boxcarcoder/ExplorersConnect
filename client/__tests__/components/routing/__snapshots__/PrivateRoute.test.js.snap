// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PrivateRoute /> component. Successfully renders the component. 1`] = `
<MemoryRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "dd5pu2",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "dd5pu2",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <PrivateRoute
      authState={
        Object {
          "isAuthenticated": true,
          "loading": false,
        }
      }
      component={[Function]}
    >
      <Route
        render={[Function]}
      >
        <mockComponent
          history={
            Object {
              "action": "POP",
              "block": [Function],
              "canGo": [Function],
              "createHref": [Function],
              "entries": Array [
                Object {
                  "hash": "",
                  "key": "dd5pu2",
                  "pathname": "/",
                  "search": "",
                  "state": undefined,
                },
              ],
              "go": [Function],
              "goBack": [Function],
              "goForward": [Function],
              "index": 0,
              "length": 1,
              "listen": [Function],
              "location": Object {
                "hash": "",
                "key": "dd5pu2",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
              "push": [Function],
              "replace": [Function],
            }
          }
          location={
            Object {
              "hash": "",
              "key": "dd5pu2",
              "pathname": "/",
              "search": "",
              "state": undefined,
            }
          }
          match={
            Object {
              "isExact": true,
              "params": Object {},
              "path": "/",
              "url": "/",
            }
          }
        >
          <div>
            Component
          </div>
        </mockComponent>
      </Route>
    </PrivateRoute>
  </Router>
</MemoryRouter>
`;
